### ISSUE TITLE
Grid Search CV .fit results don't match model .fit results when split data is exactly same 

### ISSUE THREAD

**Issue by kaurrachneet6** [skorch-dev/skorch#740]
Hi,
This is my net:
```
net = NeuralNet(
    model,
    max_epochs = 2,
    lr = .0001,
    criterion=nn.CrossEntropyLoss,
    optimizer=torch.optim.Adam,
    device = device_,
    iterator_train__shuffle=True,
    train_split = skorch.dataset.CVSplit(5, random_state = 0), 
    batch_size= -1, #Batch size = -1 means full data at once 
    callbacks=[EarlyStopping(patience = 100, lower_is_better = True, threshold=0.0001), 
    #('lr_scheduler', LRScheduler(policy=torch.optim.lr_scheduler.StepLR, step_size = 500)),
    (EpochScoring(scoring=DLutils.accuracy_score_multi_class, lower_is_better = False, on_train = True, name = "train_acc")),
    (EpochScoring(scoring=DLutils.accuracy_score_multi_class, lower_is_better = False, on_train = False, name = "valid_acc")),
    (TrainEndCheckpoint(f_params='params.pt', f_optimizer='optimizer.pt', f_criterion='criterion.pt', f_history='history.csv', f_pickle=None, fn_prefix='train_end_', dirname= self.save_path)),
#             (PrintLog(keys_ignored=None))
              ]
)
```
First, I used GridSearchCV to explore best hyperparameters as:
```
self.gkf = StratifiedGroupKFold(n_splits=5)  
GridSearchCV(self.pipe, param_grid = self.hyperparameter_grid, scoring = scores, \
                                        n_jobs = 1, cv = self.gkf.split(self.X_sl_, self.Y_sl_, groups=self.PID_sl_), \
                                        refit = False)
```
Now, once I have the best hyperparameters, I would like to fit the same best model again with same data over the 5 folds, to plot training loss over the 5 folds (because via grid search I could not retain the history for best set of parameters). So, I split the data manually in the same way as I used it in grid search CV, and then use model.fit( ) to fit the model with best parameters on each of these splits and save the history to plot the training loss over 5 folds. 
Here is how I do it:
```
for fold, (train_ix, val_ix) in enumerate(self.gkf.split(self.X_sl_, self.Y_sl_, groups=self.PID_sl_)):
    # select rows for train and test
    trainX, trainY, valX, valY = self.X_sl[train_ix], self.Y_sl[train_ix], self.X_sl[val_ix], self.Y_sl[val_ix]
    pipe_optimized.fit(trainX, trainY)
    history_fold = pd.DataFrame(pipe_optimized['net'].history)
    histories.append(history_fold)
```
I have checked that train_ix, val_ix in splitting like this and via GridSearchCV above match exactly, so I am using the same data in each fold and also same selected model with best parameters. but my results don't match. i.e. the train_loss, val_loss, train_acc and valid_acc printed while fitting is different in each fold if I do it via GridSearchCV and manually via above snippet. 
Can you help me replicate the results, or if there is some other way to retain the history of each fold for best model in Grid search CV? Thanks.

**Comment by BenjaminBossan** [comment]
>  I would like to fit the same best model again with same data over the 5 folds, to plot training loss over the 5 folds (because via grid search I could not retain the history for best set of parameters)

It depends a bit on how detailed you would like to have the result. If you're just interested in the final train scores, you can pass `return_train_score=True` to `GridSearchCV`. After fitting the `GridSearchCV` instance, access all scores via the `cv_results_` attribute.

Now if you're interested in how the train scores behave over the epochs, you can log the output of the net using a callback, e.g. like this:

```python
histories = []

from skorch.callbacks import Callback

class LogHistory(Callback):
    def on_train_end(self, net, *args, **kwargs):
        histories.append(net.history)

net = NeuralNet(..., callbacks=[..., LogHistory()])
```

As to why the results are not exactly the same: Are they close or completely different? If they are close, it could just be randomness, given how the weights are initialized randomly. Usually, you could just fix the PyTorch random seed(s), but here, it might not work. The reason is that the grid search will test a lot of different combinations of parameters, only one set of which corresponds to the best parameters. So when this parameter set is reached, your random seed has already "moved" a few places. What you would need to do is fixing the random seed again and again at the start of each model fit. I believe this should work:

```python
class FixSeed(Callback):
    def __init__(self, seed=0):
        self.seed = seed

    def initialize(self):
        torch.manual_seed(self.seed)
        torch.cuda.manual_seed(self.seed)
        return super().initialize()

net = NeuralNet(..., callbacks=[FixSeed(), ...])
```

Another remark, since this is something people sometimes confuse: The skorch-internal validation scores (the ones printed during training) do not equal the validation scores recorded by the grid search -- in fact, you might want to turn skorch-internal validation off (`NeuralNet(..., train_split=False, verbose=0)`).

**Comment by kaurrachneet6** [comment]
Thanks a lot. This solved my question.
